File: Next-Greater-Element-II.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        Arrays.fill(res, -1);  
        Stack<Integer> s = new Stack<>();
        
        for (int i = 2 * n - 1; i >= 0; i--) {
            while (!s.isEmpty() && s.peek() <= nums[i % n]) {
                s.pop();
            }
            
            if (i < n && !s.isEmpty()) {
                res[i % n] = s.peek();
            }
            
            s.push(nums[i % n]);
        }
        
        return res;
    }
}

File: Next-Greater-Element-II.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var nextGreaterElements = function(nums) {
    const n = nums.length;
    const res = new Array(n).fill(-1);  
    const stack = [];

    for (let i = 2 * n - 1; i >= 0; i--) {
        while (stack.length && stack[stack.length - 1] <= nums[i % n]) {
            stack.pop();
        }

        if (i < n && stack.length) {
            res[i % n] = stack[stack.length - 1];
        }

        stack.push(nums[i % n]);
    }

    return res;
};
